### ISSUE TITLE
Allow sessions to be enumerated

### ISSUE THREAD

**Issue by azixMcAze** [unosquare/embedio#335]
**Is your feature request related to a problem? Please describe.**
Hi, I am porting my app currently using EmbedIO 2.3 to EmbedIO 3.0 and I a have a simple helper function that lists the content of a `SessionInfo` to the console :
```csharp
SessionInfo session;
...
foreach (var kvp in session.Data)
    sb.AppendLine($"{kvp.Key}={kvp.Value}");
...
```
I try to convert it using a `ISession` parameter instead but It does not work because `ISession` does not contains a definition for `GetEnumerator`
```csharp
ISession session;
...
foreach (var kvp in session)
    sb.AppendLine($"{kvp.Key}={kvp.Value}");
...
```

**Describe the solution you'd like**
I would like to be able to enumerate a `ISession` object like a `IReadOnlyDictionary<string, object>` object.
I am not sure if it is a problem with EmbedIO's `ISession` or SWAN's `IDataDictionary`.

**Describe alternatives you've considered**
Looking at the source, it seems that `SessionImpl` uses a `DataDictionary<string, object>` that implements `IReadOnlyDictionary<TKey, TValue>`. But this does not help as `ISession` cannot be casted to this type

**Comment by rdeago** [comment]
Hello @azixMcAze, thanks for your input. 

The absence of `GetEnumerator` from `IDataDictionary<TKey,TValue>` was a deliberate choice, as stated [in XML documentation](https://github.com/unosquare/swan/blob/b48e7321d1ba89b2a4160faeb93d495816568396/src/Swan.Lite/Collections/IDataDictionary.cs#L12). Since both EmbedIO v3.0 and SWAN v2.0 are still in preview, XML docs have not yet been ported to the official documentation site, so this is no RTFM.

What's funny is that both existing implementations of `IDataDictionary<TKey,TValue>` also implement both `IDictionary<TKey,TValue>` and `IReadOnlyDictionary<TKey,TValue>`! At first, though, I was not sure I could implement them in full, especially with regard to concurrent data access. I've later discovered that it can be done, so now `IDataDictionary<TKey,TValue>` could be changed to derive from both `IDictionary<TKey,TValue>` and `IReadOnlyDictionary<TKey,TValue>`, adding only the `IsEmpty` property.

The need for `IDataDictionary<TKey,TValue>` still stands, because the main contract it defines is the assurance that `null` values will not make their way into the underlying dictionary. This makes `dict.Remove(key)` and `dict[key] = null` synonyms, for example. It also guarantees to never throw `KeyNotFoundException`, returning `null` in place of any missing value. This mimics the way session data was retrieved in EmbedIO v2, but in a more formal way, bulding the contract in the data structure itself instead of relying on an extension method (and on people consistently using it).

So far, so good. But is it a problem with EmbedIO, or with SWAN? Well, here's a piece of trivia: `IDataDictionary<TKey,TValue>` and its implementations were born in EmbedIO and later moved to SWAN. So, in a sense, it's a problem with both. 😄 SWAN needs to be modified, then EmbedIO needs to update its SWAN dependency. Stay tuned.

**Comment by rdeago** [comment]
As it turns out, no, sessions cannot be enumerated, but not for lack of functionality in `IDataDictionary<TKey,TValue>`: there is an additional lock, used for reference counting, that prevents enumeration.

So I added a `TakeSnapshot` method to `ISession`, that returns a list of key / value pairs reflecting session data as they were at the time of calling. Should be enough for debugging purposes.

```csharp
ISession session;
...
foreach (var kvp in session.TakeSnapshot())
    sb.AppendLine($"{kvp.Key}={kvp.Value}");
...
```

**Comment by geoperez** [comment]
This change was released in nuget v3.0 RC 6.

**Comment by azixMcAze** [comment]
Thank you, this is great!
